### PR TITLE
refactor: relocate CLI business logic to transport-neutral _app/ layer (ADR-0021)

### PR DESCRIPTION
**Pull request by teng-lin**
## Summary

Relocates all CLI business logic into a transport-neutral `src/notebooklm/_app/` layer so it is reusable by the CLI **and** other front-ends (a FastMCP server, a future HTTP surface) without dragging Click/Rich in. The CLI becomes a thin adapter; `_app/` is the shared neutral core. The decision is recorded in **ADR-0021**.

**Non-breaking:** underscore-private relocation; `--json` output byte-stable; existing VCR cassettes reused (zero re-records); major CLI options preserved; the `patch("cli…")` test seams intact (no command modules moved). **0 additions** to `scripts/api-compat-allowlist.json`.

## What moved

Every command area's neutral logic now lives in `_app/` (28 modules): notebooks, sources (×7), download, generate (×3), chat, labels, notes, sharing, skill, language, doctor, research, session, auth_check, profile — plus the Wave-0 foundation (`serialize`/`errors`/`resolve`/`events`). The CLI keeps `--json` envelope assembly, exit-code policy, rendering, and the interactive login flows.

### Contract

Typed frozen-dataclass `Request`/`Plan`/`Result`; a pure `build_<verb>_plan` (raises the public `notebooklm.exceptions` hierarchy); `async execute_<verb>(plan, client, *, progress) -> Result`. `_app/` imports no `click`/`rich`/`cli`/`fastmcp` (boundary lint). `_app.errors.classify` is the single neutral source of the failure-category decision each adapter projects onto its own codes — enforced by a CLI↔classify consistency gate.

## Tests

- **Direct `_app` unit coverage: 145 → 623** (every previously-untested module now has direct tests; `doctor` and `chat` had zero before).
- **Full suite: 9254 passed / 0 failed**; mypy clean; ruff clean; freshness + boundary gates green.
- **Live e2e validated:** 11 tests passed against the real NotebookLM API — notebooks (list/get/create/rename/delete/ask/summary), sources (add/rename/delete), labels (create/list/delete).
- MCP reuse proven: `_app.serialize.to_jsonable` is wire-byte-identical to the MCP server's private serializer.

## Docs

- New **ADR-0021** (transport-neutral `_app/` layer).
- CLAUDE.md slimmed 545 → 76 lines; the file map + repository tree moved into `docs/architecture.md` (which gained an `_app/` narrative + the ADR-0021 cross-ref); the freshness gate was repointed.

## Polish / review

Each domain wave was `/polish`'d (code-simplifier + claude+codex review + ultrathink) when it landed — that caught real issues (an eager-vs-lazy `get_context_path` logout-timing regression; a CLI tautology test asserting a hand-rolled filter instead of the production one). This PR additionally reviews the integration glue (the `_app/__init__` union, the CLAUDE.md/architecture.md move, the classify consistency gate).

## Deferred follow-ups

- `feat/mcp-server` rebase onto `_app/` + delete its private `_serialize`/`_ids`/`_errors` copies.
- De-monkeypatch the CLI tests via `ctx.obj` injection.

## Decided against (with rationale)

- **Resolve consolidation** — `cli/resolve.py` is a justified rich *adapter* over the pure `_app/resolve.py` core (it adds `ClickException`, console `emit_status`, `entity_name`/`list_command` hints, the `allow_full_id_passthrough` flag), not accidental duplication.
- **Test cutting** — the ≈89 % CLI/`_app` coverage overlap is layered (unit + integration), not same-layer redundant; the `--cov` gate confirms almost nothing is safely deletable.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Documentation**
  * Updated architecture docs and CLAUDE.md with a clearer file-map, concise usage/testing guidance, and common pitfalls.

* **New Features**
  * More consistent CLI behavior and JSON envelopes across commands; improved progress reporting and richer diagnostic outputs (auth/doctor/research).
  * Safer source-add validation, more robust generation/download flows, and clearer artifact/note/share/label operations.

* **Chores**
  * Core command logic centralized for consistent validation, error classification, and rendering.

* **Tests**
  * Added guardrails and consistency tests enforcing boundaries and error-category → CLI-code mapping.

### CODE REVIEW

**Comment by teng-lin** [comment]
@claude review

File: tests/unit/app/test_app_auth_check.py [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+    storage.write_text(json.dumps(_valid_storage_state("HSID", "SSID")), encoding="utf-8")
+    plan = _plan(storage_path=storage)
+
+    result = await run_auth_check(plan, read_env_auth_json=_never_read_env)
+
+    assert result.checks["cookies_present"] is True
+    assert result.checks["sid_cookie"] is True
+    assert "SID" in result.details["cookies_found"]
+    # Google-domain grouping is surfaced for the renderer.
+    assert ".google.com" in result.details["cookie_domains"]
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply tests/unit/app/test_app_auth_check.py]: Addressed in 193f5c0a: rewrote the assertion as an explicit per-element `==` over the `cookie_domains` list (it's `sorted(cookies_by_domain.keys())` — exact domain keys, never URL sanitization), so the substring `in` sink the rule keys on is gone. False positive on a test assertion; production `_app/auth_check.py` has no substring URL check.

File: src/notebooklm/_app/session.py [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+            ),
+            paths=inputs.path_info,
+            has_env_auth=inputs.has_env_auth,
+        )
+
+    return StatusReport(
+        context=StatusContext(
+            has_context=True,
+            notebook_id=inputs.notebook_id,
+            title=data.get("title"),
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply src/notebooklm/_app/session.py:209]: Addressed in c16167a1: a non-dict JSON root now routes to the same graceful `payload_readable=False` report as corrupt JSON (with the actual type logged), so `data.get(...)` can't raise `AttributeError`. I went graceful rather than raising `TypeError` because `read_status` is a read-only diagnostic that must not crash on a malformed context file — it already degrades gracefully for `JSONDecodeError`, and a non-dict root is the same failure class. Regression test `test_read_status_non_dict_context_flags_not_readable` added.

File: src/notebooklm/_app/language.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+
+    def get_config(self) -> dict:
+        """Read config from ``config.json`` (empty dict on missing/corrupt)."""
+        config_path = self._config_path()
+        if config_path.exists():
+            try:
+                return json.loads(config_path.read_text(encoding="utf-8"))
+            except json.JSONDecodeError as e:
+                logger.warning("Config file corrupted, using defaults: %s", e)
+                return {}
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply src/notebooklm/_app/language.py:166]: Addressed in c16167a1: a non-dict root is treated as corrupt per `get_config`'s documented "empty dict on missing/corrupt" contract → returns `{}` (actual type logged), so `get_language()` never `.get()`s a non-dict. Kept it graceful rather than raising `TypeError` to preserve that lenient contract. Regression test `test_non_dict_root_returns_empty_dict` added.

**Comment by teng-lin** [comment]
Thanks for the thorough review (and the strengths callouts). Addressed in `f6801c62` (+ `d6585923`):

- **#1 dead `json_output`** — removed from `resolve_source_for_delete` + `resolve_source_by_exact_title` and stopped forwarding it.
- **#4 duplicate UUID regex** — `looks_like_full_source_id` now reuses `resolve.FULL_ID_PATTERN` (single source of truth); dropped the now-unused `import re`.
- **#5 process docs** — removed `docs/scratch/` entirely and gitignored it; the load-bearing rationale lives in ADR-0021.
- **minor** — `generate_retry`: `if not result` → `if result is None` for explicit intent.

Deferred to the **MCP rebase** (a documented follow-up), per your framing that #2/#3 are MCP-facing trade-offs:
- **#2 Rich markup in `SourceIdResolution.status_message`** — documented the wrinkle in the dataclass docstring; the structural move (markup → `cli/_source_render.py`, field → plain text) will land with the MCP adapter so the CLI re-wrap + MCP plain-consumption are validated together.
- **#3 `json_output` plan-field naming** (→ `non_interactive`/`structured_output`) and the `DoctorReport`/`ProgressEvent.kind` typing notes — tracked for the MCP rebase (low priority per your note; defensible while the CLI is the only adapter).

File: src/notebooklm/cli/services/source_research.py [hunk 4]
```diff
@@ -1,206 +1,55 @@
...
-
-    outcome: SourceAddResearchOutcome
-    plan: SourceAddResearchPlan
-    start_task_id: str | None = None
-    poll_task_id: str | None = None
-    sources: list[dict[str, Any]] = field(default_factory=list)
-    report: str = ""
-    status: str | None = None
-    import_result: ResearchImportResult | None = None
+    from ....client import NotebookLMClient
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply src/notebooklm/cli/services/source_research.py:39]: Fixed in 896ef2d1: `....client` → `...client` (the 4-dot import climbed above the package root; TYPE_CHECKING-hidden so it never failed at runtime). Real latent bug — thanks.

File: src/notebooklm/_app/source_add.py [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+    if not follow_symlinks:
+        for component in [raw, *raw.parents]:
+            if component.is_symlink():
+                raise SourceAddValidationError(
+                    "Path is a symlink; pass --follow-symlinks to follow it "
+                    f"explicitly. Refusing to upload: {raw}"
+                )
+
+    file_path = raw.expanduser().resolve()
+
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply src/notebooklm/_app/source_add.py]: Fixed in 896ef2d1: expand `~` **before** the symlink check, so a `~`-prefixed symlink can no longer bypass `follow_symlinks=False`.

File: src/notebooklm/_app/chat.py [hunk 6]
```diff
@@ -0,0 +1,403 @@
...
+    A predefined ``--mode`` short-circuits to ``client.chat.set_mode`` (and the
+    result carries the mode for the adapter's envelope/prose). Otherwise the
+    persona / response-length settings are applied via ``client.chat.configure``;
+    a non-empty ``persona`` selects the ``CUSTOM`` goal. The ``goal_name`` is the
+    lowercase enum name (e.g. ``"custom"``) for a stable, human-readable JSON
+    contract.
+    """
+    if chat_mode:
+        await client.chat.set_mode(notebook_id, _MODE_MAP[chat_mode])
+        return ConfigureResult(
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply src/notebooklm/_app/chat.py:302]: Fixed in 896ef2d1: an unknown `chat_mode` now raises the public `ValidationError` instead of `KeyError`.

File: src/notebooklm/_app/download.py [hunk 7]
```diff
@@ -0,0 +1,844 @@
...
+    # Format-choice extraction. The adapter param name is data-driven via
+    # ``spec.format_param_name`` (default ``"output_format"``, slide-deck
+    # overrides to ``"slide_format"``). Leaves with no ``--format`` flag have
+    # empty ``format_choices``.
+    format_choice = ""
+    if config.format_choices:
+        format_choice = (
+            args.get(config.format_param_name, config.format_default) or config.format_default
+        )
+
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply src/notebooklm/_app/download.py:473]: Fixed in 896ef2d1: `build_download_plan` now validates the format against `format_choices` and raises `DownloadPlanValidationError` (no more silent extension fallback).

File: src/notebooklm/_app/generate_plans.py [hunk 8]
```diff
@@ -0,0 +1,753 @@
...
+        language=language,
+        wait=common["wait"],
+        timeout=common["timeout"],
+        interval=common["interval"],
+        max_retries=common["max_retries"],
+        json_output=common["json_output"],
+        params={
+            "audio_format": _AUDIO_FORMAT_MAP[raw_args["audio_format"]],
+            "audio_length": _AUDIO_LENGTH_MAP[raw_args["audio_length"]],
+        },
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply src/notebooklm/_app/generate_plans.py]: Fixed in 896ef2d1: added `_require_choice` and routed all 14 enum-map lookups through it → `GenerationPlanValidationError` instead of a raw `KeyError`.

File: src/notebooklm/_app/source_clean.py [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+    failures: list[CleanFailure] = []
+    for i in range(0, len(delete_list), chunk_size):
+        chunk = delete_list[i : i + chunk_size]
+        delete_tasks = [delete_source(notebook_id, sid) for sid in chunk]
+        results = await asyncio.gather(*delete_tasks, return_exceptions=True)
+        for sid, result in zip(chunk, results, strict=True):
+            if isinstance(result, Exception):
+                failures.append((sid, str(result)))
+            else:
+                deleted += 1
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply src/notebooklm/_app/source_clean.py:176]: Fixed in 896ef2d1: re-raise non-`Exception` `BaseException` (CancelledError / interrupts) instead of counting a cancelled delete as a success; real `Exception`s are still recorded as failures.

File: src/notebooklm/_app/auth_check.py [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+        details["error"] = read_error
+        return AuthCheckResult(plan=plan, checks=checks, details=details)
+    checks["json_valid"] = True
+
+    # Check 3: cookies present + SID lookup.
+    try:
+        cookies = extract_cookies_from_storage(storage_state)
+        checks["cookies_present"] = True
+        checks["sid_cookie"] = "SID" in cookies
+        details["cookies_found"] = list(cookies.keys())
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply src/notebooklm/_app/auth_check.py:180]: Fixed in 896ef2d1: `cookies_present = bool(cookies)` (was hardcoded `True`).

File: docs/architecture.md [hunk 11]
```diff
@@ -784,6 +799,363 @@ Vocabulary that recurs in this document and the surrounding code.
...
+| `_auth/paths.py` | Storage paths and filesystem helpers |
+| `_auth/extraction.py` | Cookie/token extraction from browser sessions |
+| `_auth/headers.py` | HTTP header construction |
+| `_auth/cookies.py` | Cookie map manipulation + `_update_cookie_input` |
+| `_auth/cookie_policy.py` | Cookie-domain allowlist and policy decisions |
+| `cli/label_cmd.py` | `label` command group (list/sources/generate/create/rename/emoji/add/delete); thin Click shells over `client.labels` and the label-listing service (ADR-0008) |
+| `cli/services/label_listing.py` | `label` CLI service: the `label list` members→source-titles join (`execute_label_list`/`LabelListPlan`). Re-exports `resolve_label_id` + `LabelResolutionError` from `_app/labels.py` (the composite `<id|name>` resolver moved to the neutral layer; the re-export keeps `from .services.label_listing import resolve_label_id` resolving for the command layer + tests) |
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply docs/architecture.md]: Fixed in 896ef2d1: escaped the pipe (`<id\|name>`) so the table row stays two columns. (The other occurrence is inside the fenced tree block, where a literal pipe is fine.)

File: src/notebooklm/_app/language.py [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+            except OSError as e:
+                logger.warning("Could not read config file: %s", e)
+                return {}
+        return {}
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply src/notebooklm/_app/language.py]: Already fixed in c16167a1: a non-dict JSON root is treated as corrupt → `{}` (with the actual type logged), per `get_config`'s documented 'empty dict on missing/corrupt' contract, so `get_language()` never `.get()`s a non-dict. Regression test added.

File: src/notebooklm/_app/session.py [hunk 13]
```diff
@@ -0,0 +1,412 @@
...
+            is_owner=data.get("is_owner"),
+            created_at=data.get("created_at"),
+            conversation_id=data.get("conversation_id"),
+        ),
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply src/notebooklm/_app/session.py:213]: Already fixed in c16167a1: a non-object JSON root routes to the same graceful `payload_readable=False` report as corrupt JSON (actual type logged), so `data.get(...)` can't raise. Regression test added.

File: src/notebooklm/_app/sharing.py [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+    return resolved_id
+
+
+async def execute_share_remove_user(
+    client: NotebookLMClient,
+    notebook_id: str,
+    email: str,
+) -> None:
+    """Remove a user's access to the notebook (raises on real failure)."""
+    await client.sharing.remove_user(notebook_id, email)
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply src/notebooklm/_app/sharing.py:127]: Holding on this one — it's correct as-is. The CLI `share remove` resolves the notebook id in its MutationPlan `resolve` callback **before** the confirm prompt and passes the resolved id to `execute_share_remove_user`; injecting a resolver here would duplicate that resolution or move it after the confirm. The neutral-API uniformity (an MCP caller passing a partial id to remove, like the other share flows) is real but is MCP-rebase work — tracked there alongside the other MCP-facing adapter changes.

File: docs/scratch/2026-06-08-cli-refactor-log.md [hunk 15]
```diff
@@ -0,0 +1,332 @@
...
+## CLAUDE.md slimmed (user: "keep it very slim!!!") — 545 → 76 lines.
+The `## Architecture` section (Key Files table + Repository Structure tree, ~353 lines) was redundant with the
+narrative `docs/architecture.md` (which had NO file map and NO `_app/` mention). Consolidated: moved the file map
+into `docs/architecture.md` as a new `## File map` section (keeping the `### Repository Structure` heading the
+freshness gate parses), added a `_app/` narrative section + ADR-0021 cross-ref + fixed its "See also → CLAUDE.md
+file index" pointer. Repointed `check_claude_md_freshness.py` (default `--claude-md` → docs/architecture.md), its
+real-file test, and the CI step at it. Slimmed the rest of CLAUDE.md to essentials (overview + critical RPC
+constraint, commands, before-pushing, architecture pointer, pitfalls, usage, testing, docs index, PR workflow),
+preserving the install-docs-gated `uv sync` block + installation link. Gates green: freshness (architecture.md),
+docs-module-refs (the moved table's inline refs all resolve; `_core.py` allowlist now genuinely used), install-docs,
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply docs/scratch/2026-06-08-cli-refactor-log.md]: Resolved — `docs/scratch/` was removed entirely in d6585923 (process notes; the load-bearing rationale lives in ADR-0021).

File: docs/scratch/2026-06-08-cli-refactor-plan.md [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+
+## 1. Decision (unchanged — both oracles + both momus endorse it)
+
+Relocate transport-neutral **business logic** into `src/notebooklm/_app/` (underscore-private,
+ADR-0012). CLI / MCP / future HTTP are thin sibling adapters. `cli/services/*` shrink to adapters;
+no `usecases/` ceremony, no DI container. Supersedes the earlier in-place-utilities proposal on *placement*.
+
+```
+client.* (public domain API) → _app/ (neutral) → cli/(Click) + mcp/(FastMCP) [+ http/]
+```
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply docs/scratch/2026-06-08-cli-refactor-plan.md]: Resolved — `docs/scratch/` removed in d6585923.

File: docs/scratch/2026-06-08-cli-refactor-plan.md [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+| Prototype command | ID resolution today | Risk |
+|---|---|---|
+| `download *` | inside `execute_download` (download.py:627), full-ID fast-path kept (resolve.py:285) | **LOW** (verified) — no new preflight |
+| `generate *` | inside `execute_generation` (generate.py:180) | **LOW** (verified) |
+| `source clean` | notebook resolve in command; no per-source id | **LOW** (no new RPC) |
+| `source add-research` | notebook resolve in command; full-ID fast path | **MED** — keep resolve order; assert no extra GET |
+
+Rule: moving resolution between layers must NOT add a list/get preflight on a full-ID path. Each wave
+runs **all 15** `tests/integration/cli_vcr/` suites.
+
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply docs/scratch/2026-06-08-cli-refactor-plan.md]: Resolved — `docs/scratch/` removed in d6585923.

**Comment by teng-lin** [comment]
@claude review

**Comment by teng-lin** [comment]
Thanks for the thorough re-review and for confirming all 14 prior fixes.

**New finding — fixed in `5af0f460`:** good catch, I'd missed the sibling lookup. `_RESPONSE_LENGTH_MAP[response_length]` in `execute_configure` now uses the same `try/except KeyError → ValidationError` guard already applied to `_MODE_MAP[chat_mode]` in that function (grepped chat.py — those were the only two direct `_MAP[...]` lookups).

**Held items** — agreed on all three (share-remove resolution / `SourceIdResolution` markup / `json_output` naming); tracked for the MCP rebase.

**Minors:**
- `check_claude_md_freshness.py` residual "CLAUDE.md" exit-code/error prose → reworded to "structure doc" in `5af0f460` (the default target is `docs/architecture.md` now).
- `DoctorReport.checks` TypedDict + `ProgressEvent.kind` `Literal`/`Enum` — agreed these would help future adapters; low-priority typing polish, tracking for a follow-up rather than expanding this PR.
